<a href="https://github.com/leightonpayne/padloc/LICENSE" alt="License"><img src="https://img.shields.io/github/license/leightonpayne/padloc" /></a> <a href="https://github.com/leightonpayne/padloc/" alt="Last update"><img src="https://img.shields.io/github/last-commit/leightonpayne/padloc?label=last%20update" /></a> <a href="https://github.com/leightonpayne/padlocDB/releases" alt="Release"><img src="https://img.shields.io/github/v/release/leightonpayne/padlocDB" /></a> 

# PADLOC: Prokaryotic Antiviral Defence LOCator

### About

PADLOC is a software tool for identifying antiviral defence systems in prokaryotic genomes. PADLOC screens against a curated database to annotate defence systems based on sequence homology and genetic architecture.

You can also use PADLOC via our [web app](https://padlocbio.com).

### Citation

tbd

### Installation

PADLOC can be installed by cloning or downloading this github repository and running the setup script.

1. Clone repo:

```bash
git clone https://github.com/leightonpayne/padloc $HOME/padloc
```

2. Add to `$PATH`

```bash
export PATH="$HOME/padloc:$PATH"
```

2. Run setup:

```bash
padloc --setup
```

The setup script installs software dependencies via [Homebrew](https://brew.sh/) and R packages from [CRAN](https://cran.r-project.org/), if they are not already installed. It also downloads and compiles the default database used by PADLOC from [PADLOC-DB](https://github.com/leightonpayne/padloc-db).

### Usage

#### Examples

```
# 
$ padloc --faa genome.faa --gff features.faa
```



```bash
# 
# Speed up runtime with multiple cpus and save output to mydir 
padloc --faa --gff --outdir mydir --cpu 4

```

#### Output

```
<outdir>
├── domtblout
│   └── *.domtblout         <- Domain table files generated by HMMER.
├── prodigal
│   ├──
│   └── *.gff         <- Domain table files generated by HMMER.
├── *_pdlcout.csv           <- padloc output file.
└── *.gff                   <- GFF annotation file for identified defence systems.
```

If the same output directory is used when processing multiple genomes, `*.domtblout` files will accumulate in the `domtblout/` directory. If the `*.domtblout` file already exists for a particular genome, `hmmsearch` will not be run again for that genome unless `--append` is specified. In which case, the output of `hmmsearch` will be appended to the existing `*.domtblout` file. When processing multiple genomes in batches, this allows for the pipeline to be resumed if it is interrupted.



### Command line options

```
Required:
  --faa <f>      Path to amino acid fasta file
General:
  --help         Print help information
  --version      Print version information
  --citation     Print citation information
  --debug        Run with debug information
Setup:
  --setup        Set up a fresh padloc install
  --bootstrap    Install mandatory dependencies (called by --setup)
  --updatedb     Install latest database (called by --setup)
Input:
  --gff <f>      Use the gff file <f> for previous annotations
  --data <d>     Use the directory <d> as an alternative data directory
Output:
  --outdir <d>   Path to the output directory <d>
  --rawout       Include a summarised raw output file for the genome searched
Optional:
  --cpu <n>      Set number of cpus to <n>, the default is 1
  --append       Append the hmmer output to exisitng results in domtblout/
  --force        Write over hmmer output in domtblout/
```

### PADLOC-DB

The database of HMMs and defence system models used by PADLOC is available at the repository: [PADLOC-DB](https://github.com/leightonpayne/padloc-db). A custom database can be specified with `--data [d]`, refer to [PADLOC-DB](https://github.com/leightonpayne/padloc-db) for more information about configuring a custom database.

### Issues

Bugs and feature requests can be submitted to the [Issues tab](https://github.com/leightonpayne/padloc/issues).

### Dependencies

These dependencies are installed automatically during setup.

### Mandatory

- **R**  
  *R Core Team (2018). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. https://www.R-project.org/.*
  - **tidyverse**  
    *Wickham et al. (2019). Welcome to the tidyverse. Journal of Open Source Software, 4(43), 1686, https://doi.org/10.21105/joss.01686.*
  - **yaml**  
    *Stephens, J., et al. (2020). yaml: Methods to Convert R Data to YAML and Back. https://CRAN.R-project.org/package=yaml.*
  - **getopt**  
    *Davis, T., et al. (2019). getopt: C-Like 'getopt' Behavior. https://CRAN.R-project.org/package=getopt.*
- **HMMER3**  
*Finn, R.D., Clements, J., and Eddy, S.R. (2011). HMMER web server: interactive sequence similarity searching. Nucleic Acids Res 39, W29–W37.*

### Optional

- **Prodigal**  
  *Hyatt, D., Chen, GL., Locascio, P.F., Land, M.L., Larimer, F.W., and Hauser, L.J. (2010) Prodigal: prokaryotic gene recognition and translation initiation site identification. BMC Bioinformatics 11, 119.*

## License

This software and data is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).